File: sdk/ruby/lib/rocksky/http.rb
require "json"
require "net/http"
require "uri"

module Rocksky
  # Low-level XRPC transport. Most users should go through the resource
  # accessors on {Rocksky::Client} (e.g. `client.actor.get_profile(...)`).
  class HTTP
    DEFAULT_OPEN_TIMEOUT = 10
    DEFAULT_READ_TIMEOUT = 30

    attr_reader :client

    def initialize(client)
      @client = client
    end

    # GET /xrpc/<nsid>?...
    def query(nsid, params = {})
      request(method: :get, nsid: nsid, params: params)
    end

    # POST /xrpc/<nsid>?... with optional JSON body.
    def procedure(nsid, params = {}, body = nil)
      request(method: :post, nsid: nsid, params: params, body: body)
    end

    private

    def request(method:, nsid:, params: {}, body: nil)
      uri = build_uri(nsid, params)
      req = build_request(method, uri, body)
      apply_headers(req)

      response = perform(uri, req)
      handle_response(response, nsid)
    rescue Timeout::Error, Errno::ECONNREFUSED, SocketError, IOError,
           Net::OpenTimeout, Net::ReadTimeout => e
      raise Error.from_transport(e, nsid: nsid)
    end

    def build_uri(nsid, params)
      url = "#{client.base_url}/xrpc/#{nsid}"
      uri = URI.parse(url)
      encoded = encode_params(params)
      uri.query = URI.encode_www_form(encoded) unless encoded.empty?
      uri
    end

    def build_request(method, uri, body)
      req =
        case method
        when :get  then Net::HTTP::Get.new(uri.request_uri)
        when :post then Net::HTTP::Post.new(uri.request_uri)
        else raise ArgumentError, "unsupported method: #{method}"
        end

      if body && !body.empty?
        req["content-type"] = "application/json"
        req.body = JSON.generate(body)
      end

      req
    end

    def apply_headers(req)
      req["accept"] = "application/json"
      req["user-agent"] = client.user_agent if client.user_agent
      req["authorization"] = "Bearer #{client.token}" if client.token
      client.headers.each { |name, value| req[name] = value }
    end

    def perform(uri, req)
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = (uri.scheme == "https")
      http.open_timeout = client.open_timeout
      http.read_timeout = client.read_timeout
      http.request(req)
    end

    def handle_response(response, nsid)
      status = response.code.to_i
      body = parse_body(response)

      return body if status.between?(200, 299)

      raise Error.from_response(status, body, nsid: nsid)
    end

    def parse_body(response)
      return nil if response.body.nil? || response.body.empty?

      content_type = response["content-type"].to_s
      if content_type.include?("application/json")
        JSON.parse(response.body)
      else
        response.body
      end
    rescue JSON::ParserError
      response.body
    end

    # Drop nil values; convert symbol keys to strings; join arrays with commas
    # (matches the lexicon-defined array encoding used by other SDKs).
    def encode_params(params)
      return {} if params.nil?

      Hash(params).each_with_object({}) do |(key, value), out|
        next if value.nil?

        out[key.to_s] = encode_value(value)
      end
    end

    def encode_value(value)
      case value
      when Array then value.join(",")
      when true, false then value.to_s
      else value
      end
    end
  end
end

File: sdk/elixir/lib/rocksky/http.ex
defmodule Rocksky.HTTP do
  @moduledoc """
  Low-level XRPC transport. Most users should call the namespace modules
  (`Rocksky.Actor`, `Rocksky.Scrobble`, etc.) instead.

  All XRPC endpoints live under `/xrpc/<nsid>`. Queries are `GET` with query
  parameters; procedures are `POST` with a JSON body.
  """

  alias Rocksky.{Client, Error}

  @type nsid :: String.t()
  @type params :: keyword() | map()

  @doc """
  Issue a query (GET) against `nsid` with the given params.
  """
  @spec query(Client.t(), nsid(), params()) :: {:ok, term()} | {:error, Error.t()}
  def query(%Client{} = client, nsid, params \\ []) do
    request(client,
      method: :get,
      nsid: nsid,
      params: encode_params(params)
    )
  end

  @doc """
  Issue a procedure (POST) against `nsid`.

  `body` is encoded as JSON. `params` are sent as query string parameters
  (some lexicons declare both, e.g. `app.rocksky.player.next`).
  """
  @spec procedure(Client.t(), nsid(), params(), term()) ::
          {:ok, term()} | {:error, Error.t()}
  def procedure(%Client{} = client, nsid, params \\ [], body \\ nil) do
    request(client,
      method: :post,
      nsid: nsid,
      params: encode_params(params),
      body: body
    )
  end

  defp request(%Client{} = client, opts) do
    nsid = Keyword.fetch!(opts, :nsid)
    method = Keyword.fetch!(opts, :method)
    url = client.base_url <> "/xrpc/" <> nsid

    req_opts =
      [
        method: method,
        url: url,
        headers: build_headers(client),
        params: Keyword.get(opts, :params, []),
        decode_json: [keys: :strings],
        retry: :safe_transient
      ]
      |> maybe_put_body(Keyword.get(opts, :body))
      |> Keyword.merge(client.req_options)

    case Req.request(req_opts) do
      {:ok, %Req.Response{status: status, body: body}} when status in 200..299 ->
        {:ok, body}

      {:ok, %Req.Response{status: status, body: body}} ->
        {:error, Error.from_status(status, body)}

      {:error, reason} ->
        {:error, Error.from_transport(reason)}
    end
  end

  defp maybe_put_body(opts, nil), do: opts
  defp maybe_put_body(opts, body), do: Keyword.put(opts, :json, body)

  defp build_headers(%Client{token: nil, headers: extra}) do
    [{"accept", "application/json"} | extra]
  end

  defp build_headers(%Client{token: token, headers: extra}) do
    [
      {"authorization", "Bearer " <> token},
      {"accept", "application/json"}
      | extra
    ]
  end

  @doc false
  # Drop nil values, convert atom keys to strings, leave the rest to Req's encoder.
  def encode_params(params) when is_list(params) or is_map(params) do
    params
    |> Enum.reject(fn {_k, v} -> is_nil(v) end)
    |> Enum.map(fn {k, v} -> {to_string(k), encode_value(v)} end)
  end

  defp encode_value(v) when is_list(v), do: Enum.join(v, ",")
  defp encode_value(v), do: v
end
